### PR TITLE
Http proxy forwards incoming requests to service manager

### PR DESCRIPTION
**Pull request by fabianschwarzfritz**
Added a simple go server to the service-broker-proxy forwarding incoming http requests to the service manager.

Credentials of the service-manager are added to the kubernetes deployment as a secret, other service-broker-proxy configuration is configured via a config map. The readme now also contains a short description on how to create the config map and secrets to deploy the broker proxy on a kubernetes cluster.

### CODE REVIEW

File: yaml/service-broker-proxy-deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+          valueFrom:
+            configMapKeyRef:
+              name: service-broker-proxy-configuration
+              key: service_manager_url
+        volumeMounts:
+        - name: service-manager-secrets
+          mountPath: "/etc/service-manager-secrets"
+          readOnly: true
+      imagePullSecrets:
+      - name: service-broker-proxy-image-secret
```
Comment (mvladev) [inline yaml/service-broker-proxy-deployment.yaml]: Why is this needed? The image should be pushed to publicly available repository.

File: yaml/service-broker-proxy-deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,43 @@
+---
+apiVersion: apps/v1beta1
+kind: Deployment
+metadata:
+  name: service-broker-proxy
+spec:
+  replicas: 1
```
Comment (mvladev) [inline yaml/service-broker-proxy-deployment.yaml]: Add label selector

```yaml
  selector:
    matchLabels:
      app: service-proker-proxy
```

File: proxy/proxy.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+	//TODO error handling
+	client := &http.Client{
+		Timeout: time.Duration(proxy.config.timeoutSeconds) * time.Second,
+	}
+	smRequest, _ := http.NewRequest(request.Method, brokerURL, request.Body)
+	proxy.copyAndFilterHeaders(request, smRequest)
+	proxy.setBrokerHeaders(smRequest)
+
+	printRequest(smRequest)
```
Comment (mvladev) [inline proxy/proxy.go]: Use https://golang.org/pkg/net/http/#RoundTripper for logging

File: yaml/service-broker-proxy-deployment.yaml (hunk 1 above)
Comment (fabianschwarzfritz) [reply yaml/service-broker-proxy-deployment.yaml]: True! Thanks, I've removed the secret.

File: yaml/service-broker-proxy-deployment.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (fabianschwarzfritz) [reply yaml/service-broker-proxy-deployment.yaml]: Thanks! I have removed the *replicas* part.

Comment (fabianschwarzfritz) [reply yaml/service-broker-proxy-deployment.yaml]: I have added the selector 👍

File: proxy/proxy.go (hunk 3 above)
Comment (fabianschwarzfritz) [reply proxy/proxy.go]: I have added the RoundTripper for logging the requests and responses.

**Comment by mvladev** [comment]
I would suggest using the Kubernetes apimachinery [proxy package](https://github.com/kubernetes/apimachinery/tree/kubernetes-1.10.0/pkg/util/proxy). It supports HTTPSv2 upgrade and automatically sets the correct `X-Forwarded-For`, `X-Forwarded-Host` and `X-Forwarded-Proto` headers.

[Implementation](https://github.com/kubernetes/kubernetes/blob/fc32d2f3698e36b93322a3465f63a14e9f0eaead/pkg/kubectl/proxy/proxy_server.go#L183-L221) of said library is used for the `kubectl proxy` command.

At the end both [client-go](https://github.com/kubernetes/client-go/) and [service catalog's client](https://github.com/kubernetes-incubator/service-catalog/tree/master/pkg/client) are going to be used for this proxy and they both require the `apimachinery` library.

**Comment by fabianschwarzfritz** [comment]
Thanks for the suggestion @mvladev to use the apimachinery proxy package. I know the k8s packages have quite some code quality. I'd prefer to go the "easy route" first and simply start with the build in go http functionality. We can still iterate on this. 

Even in the beginning I'd prefer using the standard build in go lib. I assume it is easier to gain the knowledge of the standard go lib. 
Secondly, the number of http headers is quite little. I'd rather have full control of what headers are set, rather than have a proxy library set headers automatically. 

As said, this is  something we still can iterate on later and switch to the *proxy package* if we think it's beneficial.

File: yaml/service-broker-proxy-deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    metadata:
+      labels:
+        app: service-broker-proxy
+    spec:
+      containers:
+      - name: service-broker-proxy
+        image: ${REPOSITORY}/${PROJECT}:${VERSION}
+        ports:
+        - containerPort: 8080
+        env:
```
[the remark this answers is not in the document]
Comment (fabianschwarzfritz) [reply yaml/service-broker-proxy-deployment.yaml]: Yes, I totally see the advantage of mounting these configuration files. 
However the namespace cannot be changed anyhow without redeploying the proxy. So with this env (namespace) it shouldn't be a problem.
I have moved the other configurations to config maps. Also this is something I'd iterate on if needed.
